paper.md

```diff
@@ -38,89 +38,140 @@ affiliations:
+
+Once the question is answered using the widget, it works in the same way as the standard integration question regarding how it is saved. We save the name of the project retrieved from DAISY and the url to the project so it can be used for further processing.
 
-*TODO: Jan - screenshot of the DSW side, code snippet*
+![Integration widget question in DSW -- empty and with selected answer from DAISY](figures/dsw-daisy-question.png)
 
 ## Widget for DAISY Projects
 
-*TODO: Vilém (and Jan) - describe implementation on DAISY side*
+The widget implemented on DAISY's side allows users to select one of existing projects and send them back as an answer to DSW. When the widget is opened in a popup window, user is asked to log into DAISY. As the initial page is actual DAISY login form, user needs to use the DAISY login credentials. Once logged in, user is redirected to `/integrations/dsw/list-projects` DAISY endpoint presenting a list of projects he has access to. The list contains only records for which the user is appointed to as local custodian or personnel. The internal DAISY project ID, project's acronym and full title is presented. When a project of interest is clicked on, the popup window is closed, and the information is set in DSW.
 
-*TODO: Jan - screenshot of the "widget"*
+![Integration widget in DAISY showing a list of user's projects](figures/daisy-integration-widget.png)
 
 ## Retrieving Data for Document Generation
 
-*TODO: Marek - describe HTTP requests from DSW*
+To allow retrieving project data from DAISY using its REST API, we had to allow making HTTP requests from Jinja2 templates. It has been added as a new feature to the [DSW Document Worker](https://github.com/ds-wizard/document-worker) component. As it is written in [Python](https://www.python.org) programming language, we decided to use the well-known [Requests library](https://docs.python-requests.org/en/latest/). Because any data steward can upload own template and creating requests might be harmful, we also designed a configuration for the document worker that can enable this feature for specific templates (based on their IDs), set limit and timeout for the HTTP requests made from Jinja2.
+
+To make limits and timeouts effective, we had to introduce a simple wrapper class around requests library. An object of such class is then passed as variable into a Jinja2 template if the feature is enabled. On each requests through the object, a counter is increased and check against the limit. The configured timeout is plainly passed to the call of Requests library functions. It does not support any sessions as it is expected that just a single request (or few consecutive requests) will be made in behalf of document generation speed. A HTTP requests can be made from a Jinja2 template in the way shown in Figure 5.
```

Comment:
To make limits and timeouts effective, we had to introduce a simple wrapper class around the requests library. An object of such a class is then passed as a variable into a Jinja2 template if the feature is enabled. On each request through the object, a counter is increased and checked against the limit. The configured timeout is passed to the call of Requests library functions. It does not support sessions as it is expected that just a single request (or few consecutive requests) will be made to optimise document generation speed. A HTTP request can be made from a Jinja2 template in the way shown in Figure 5.

Reply:
```suggestion
To make limits and timeouts effective, we had to introduce a simple wrapper class around the requests library. An object of such a class is then passed as a variable into a Jinja2 template if the feature is enabled. On each request through the object, a counter is increased and checked against the limit. The configured timeout is passed to the call of Requests library functions. It does not support sessions as it is expected that just a single request (or few consecutive requests) will be made to optimise document generation speed. A HTTP request can be made from a Jinja2 template in the way shown in Figure 5.
```